cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -233,6 +235,43 @@ struct DivideChecked {
   }
 };
 
+struct Negate {
+  template <typename T, typename Arg0>
+  // NOTE [EPM]: Discuss on 0 vs. -0.
+  static constexpr enable_if_floating_point<T> Call(KernelContext*, Arg0 arg) {
+    return -arg;
+  }
+
+  // NOTE [EPM]: How to handle unsigned integers?
+  //  * Promote to signed?
+  //  * State that unsigned numbers are not supported (i.e., undefined behavior)?
+  //  * Use C++ integral conversions (e.g., Negate(-128) = -128)?
+  //    * https://timsong-cpp.github.io/cppwp/n4659/conv.integral
+  template <typename T, typename Arg0>
+  static constexpr enable_if_integer<T> Call(KernelContext*, Arg0 arg) {
+    return -arg;
```

Comment:
To avoid undefined behavior warnings in the overflow case, use bit ops instead of `-`
```suggestion
    return static_cast<T>(to_unsigned(~arg) + 1);
```